### PR TITLE
feat: added retry on files sync error

### PR DESCRIPTION
**Pull request by idsulik**
**Description**

Hi! I've been using skaffold for the last 2 years and from time to time I see error - "Skipping deploy due to sync error" and you can't do nothing with it in it was executed in auto sync mode.  
Within this PR I added retry mechanism if sync returns error

### CODE REVIEW

**Comment by idsulik** [comment]
Hi!
Should I do anything else or I just need to wait?

**Comment by idsulik** [comment]
@ericzzzzzzz Hi!
Should I do anything else or I just need to wait?

**Comment by ericzzzzzzz** [comment]
Hi @idsulik  I think this needs rebase, also do we know what errors are retryable and what not?  I'm thinking we should distinguish those if we implement a retry mechanism for sync error

**Comment by idsulik** [comment]
@ericzzzzzzz I think we should retry sync regardless of the error because in the case of auto-sync mode, you can't force it to sync again. there are several cases:  

- pre/post hooks error. 
- copying/deleting docker files. 


So if something goes wrong it can retry it because otherwise I have to use a workaround - checkout a different branch, checkout back to force it to sync changes

**Comment by idsulik** [comment]
do you know any error related to the sync when we shouldn't retry ?

**Comment by ericzzzzzzz** [comment]
@idsulik  You can find a few errors which are not recoverable without manual work using the query in `"sync error" in:title`.

**Comment by idsulik** [comment]
@ericzzzzzzz what do you mean?
Search for "sync error" in the project gives me only 3 lines:  
image

I've already specified these errors above.  Could you please specify 1 sync error that is not recoverable without manual action?

File: pkg/skaffold/runner/dev.go [hunk 1]
```diff
@@ -99,10 +104,23 @@ func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
...
+			return nil
+		}
+		for _, s := range r.changeSet.NeedsResync() {
+			err := backoff.Retry(
+				func() error {
+					return syncHandler(s)
+				}, backoff.WithContext(opts, childCtx),
+			)
+
+			if err != nil {
```
Comment (ericzzzzzzz) [inline pkg/skaffold/runner/dev.go:121]: why returning nil here?

File: pkg/skaffold/runner/dev.go [hunk 2]
```diff
@@ -85,9 +86,13 @@ func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
 			r.changeSet.ResetSync()
 			r.intents.ResetSync()
 		}()
+
+		// todo: make this configurable
```
Comment (ericzzzzzzz) [inline pkg/skaffold/runner/dev.go:90]: Is this something you left here while you're coding and wanted to add before committing your changes ?

File: pkg/skaffold/runner/dev.go (hunk 1 above)
Comment (idsulik) [reply pkg/skaffold/runner/dev.go:121]: it's old logic, I left it as is. 
image

File: pkg/skaffold/runner/dev.go (hunk 2 above)
Comment (idsulik) [reply pkg/skaffold/runner/dev.go:90]: No, not in this PR, I left it in case if anyone wants to do it, maybe I'll do it by myself, but not now

**Comment by ericzzzzzzz** [comment]
> @ericzzzzzzz I think we should retry sync regardless of the error because in the case of auto-sync mode, you can't force it to sync again. there are several cases:
> 
> * pre/post hooks error.
> * copying/deleting docker files.
> 
> So if something goes wrong it can retry it because otherwise I have to use a workaround - checkout a different branch, checkout back to force it to sync changes

> * pre/post hooks error.
 Do you mean hooks error will cause sync fail and we should re-try sync in this case? 
 Cloud you give me a specific example why re-try sync would help ?
 
 > * copying/deleting docker files.
  Wouldn't this trigger a rebuild?

File: pkg/skaffold/runner/dev.go (hunk 2 above)
Comment (ericzzzzzzz) [reply pkg/skaffold/runner/dev.go:90]: I'd suggest to implement configurable re-try for sync and default it to `0` -- not re-try to keep the user experience the same as before, if we want to have a sync retry, but I am still not convinced that we need this. 

While `// todo` comments might seem like helpful reminders, they often create more problems than they solve:

`Limited Visibility`: Only those who stumble upon the specific code section see the comment, potentially leaving others unaware of the task.
`Unclear Ownership and Status`: There's no indication of who's responsible for the task, whether it's assigned, in progress, or forgotten.
`Lack of Discussion`: Meaningful conversations are difficult without a dedicated issue or pull request, hindering collaboration.
`Prioritization Challenges`: Maintainers struggle to assess the urgency and importance of tasks scattered throughout the codebase.
`Ambiguity in Pull Requests`: It's unclear if // todo marks unfinished work or instructions for future changes, causing confusion during reviews.
These issues often lead to // todo items being overlooked or forgotten.

A more effective approach is to create an issue for each task:

- Issues provide a centralized location for tracking, assigning, and discussing tasks.
- They allow for clear prioritization and status updates.
- Discussions within issues create better communication and collaboration.

**Comment by idsulik** [comment]
> * Cloud you give me a specific example why re-try sync would help ?

while skaffold is trying to sync files something happened with the connection, skaffold will fail and when the connection become stable, you have to somehow trigger skaffold to start syncing again. if we have retry, on second/third retry it can sync the files successfully .  

I have such cases a lot(VPN issue, quick checkout between branches lead to sync error etc.)

**Comment by ericzzzzzzz** [comment]
> > * Cloud you give me a specific example why re-try sync would help ?
> 
> while skaffold is trying to sync files something happened with the connection, skaffold will fail and when the connection become stable, you have to somehow trigger skaffold to start syncing again. if we have retry, on second/third retry it can sync the files successfully .
> 
> I have such cases a lot(VPN issue, quick checkout between branches lead to sync error etc.)

@idsulik thank you for the explanation! Could you please fix the ci issue. I'll approve the change.

**Comment by idsulik** [comment]
@ericzzzzzzz pushed the fix. thank you!

**Review by ericzzzzzzz — APPROVED** [review]

**Pull request merged**